java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -333,23 +391,60 @@ public static SlidingWindowState create(
 
             return new SlidingWindowState(
                     aggregationFieldsDescriptor,
-                    timestampToRow,
+                    timestampToRows,
                     listState,
                     outFieldNameToLeftTimestampIdx,
                     maxRegisteredTimer,
+                    maxTriggeredTimer,
                     lastOutputRow,
                     accumulatorStates);
         }
 
         /**
-         * Add the row to the state.
+         * Add the row to the state and adjust left idx accordingly.
          *
          * @param timestamp The row time of the row.
-         * @param row The row to be added
+         * @param row The row to be added.
+         * @param affectedAggFieldIndices The indices of the agg fields that would add or have added
+         *     this row to their accumulator. If this value is null, it means that all agg fields
+         *     add this row.
          */
-        public void addRow(long timestamp, Row row) throws Exception {
-            timestampToRow.put(timestamp, row);
-            timestampList.add(timestamp);
+        public void addRow(long timestamp, Row row, Set<Integer> affectedAggFieldIndices)
+                throws Exception {
+            if (timestampToRows.contains(timestamp)) {
+                List<Row> rows = timestampToRows.get(timestamp);
+                rows.add(row);
+                timestampToRows.put(timestamp, rows);
+            } else {
+                timestampToRows.put(timestamp, new ArrayList<>(Collections.singletonList(row)));
+
+                List<Long> timestamps = getTimestampList();
+
+                if (timestamps.isEmpty() || timestamps.get(timestamps.size() - 1) < timestamp) {
+                    timestampList.add(timestamp);
+                } else {
+                    for (int i = 0; i < timestamps.size(); i++) {
```

Comment:
Let's add a TODO to use skip list for timestamps so that late data can be handled in O(n) time.